### PR TITLE
FUSETOOLS-2032 - restrict reconnection possibilities / validate new c…

### PR DESCRIPTION
**Pull request by lhein**
…onnection

### CODE REVIEW

File: editor/plugins/org.fusesource.ide.camel.editor/src/org/fusesource/ide/camel/editor/features/misc/ReconnectNodesFeature.java [hunk 1]
```diff
@@ -57,14 +53,41 @@ public void postReconnect(IReconnectionContext context) {
...
+			}
+		}
+		return null;
+	}
+	
+	private boolean isReconnectValid(IReconnectionContext context) {
+		AbstractCamelModelElement source = getNode(context.getConnection().getStart());
+		AbstractCamelModelElement newTarget = getNode(context.getNewAnchor());
+
+		return 	source != null &&
```
Comment (apupier) [inline editor/plugins/org.fusesource.ide.camel.editor/src/org/fusesource/ide/camel/editor/features/misc/ReconnectNodesFeature.java]: miss the case of trying to reconnect the source to the first element, even if it is not working after, we don't have the icon which give feedback for the forbidden action

image

**Comment by apupier** [comment]
unit tests will be great here

**Comment by apupier** [comment]
now I can't reconnect bean1 to bean2:
image

File: editor/tests/org.fusesource.ide.camel.editor.tests.integration/src/main/java/org/fusesource/ide/camel/editor/integration/CamelEditorIT.java [hunk 2]
```diff
@@ -307,6 +308,85 @@ public void dropRouteOnRoute() throws Exception {
...
+		
+		AbstractCamelModelElement createdBean = model.findNode("_bean1");
+		assertThat(createdBean).isNotNull();
+
+		// now check if the reconnect is possible - it should be
+		assertThat(ReconnectNodesFeature.canElementsConnect(inbox, createdBean)).isTrue();
+	}
+
+	@Test
+	public void reconnectToWiredNode() throws Exception {
```
Comment (apupier) [inline editor/tests/org.fusesource.ide.camel.editor.tests.integration/src/main/java/org/fusesource/ide/camel/editor/integration/CamelEditorIT.java]: better to provide more explicit test name: something like: reconnectToWiredNodeShouldNotBepossible

File: editor/tests/org.fusesource.ide.camel.editor.tests.integration/src/main/java/org/fusesource/ide/camel/editor/integration/CamelEditorIT.java [hunk 3]
```diff
@@ -307,6 +308,85 @@ public void dropRouteOnRoute() throws Exception {
 		assertThat(createRouteFigureFeature.canExecute(createCtx)).isFalse();
 	}
 	
+	@Test
+	public void reconnectToUnwiredNode() throws Exception {
```
Comment (apupier) [inline editor/tests/org.fusesource.ide.camel.editor.tests.integration/src/main/java/org/fusesource/ide/camel/editor/integration/CamelEditorIT.java]: better to provide more explicit test name: something like: reconnectToUniredNodeShouldBePossible

File: editor/plugins/org.fusesource.ide.camel.editor/src/org/fusesource/ide/camel/editor/features/misc/ReconnectNodesFeature.java [hunk 4]
```diff
@@ -51,5 +63,44 @@ public void postReconnect(IReconnectionContext context) {
 			oldTarget.setOutputElement(null);
 		}
 		CamelElementConnection f = new CamelElementConnection(source, newTarget);
```
Comment (apupier) [inline editor/plugins/org.fusesource.ide.camel.editor/src/org/fusesource/ide/camel/editor/features/misc/ReconnectNodesFeature.java]: the f variable is not used afterwise, even if this code has not been modified, it seems strange, do we need to call the constructor at all? At least we should not assign it to avairable

**Comment by apupier** [comment]
I cannot reconnect the source anchor from controlbus to Bean:
image

**Comment by apupier** [comment]
I cannot reconnect source anchor to cxf neither:
image

When drop There is this stacktrace: !ENTRY org.eclipse.graphiti.ui 4 0 2016-08-17 16:48:26.829
!MESSAGE GFCommandStack.execute(Command) java.lang.IllegalArgumentException
!STACK 0
java.lang.IllegalArgumentException
	at org.fusesource.ide.camel.model.service.core.model.CamelElementConnection.reconnect(CamelElementConnection.java:151)
	at org.fusesource.ide.camel.model.service.core.model.CamelElementConnection.<init>(CamelElementConnection.java:53)
	at org.fusesource.ide.camel.editor.features.misc.ReconnectNodesFeature.postReconnect(ReconnectNodesFeature.java:65)
	at org.eclipse.graphiti.features.impl.DefaultReconnectionFeature.reconnect(DefaultReconnectionFeature.java:119)
	at org.eclipse.graphiti.features.impl.DefaultReconnectionFeature.execute(DefaultReconnectionFeature.java:181)
	at org.eclipse.graphiti.internal.command.GenericFeatureCommandWithContext.execute(GenericFeatureCommandWithContext.java:64)
	at org.eclipse.graphiti.internal.command.GFPreparableCommand.doExecute(GFPreparableCommand.java:37)
	at org.eclipse.emf.transaction.RecordingCommand.execute(RecordingCommand.java:135)
	at org.eclipse.graphiti.ui.internal.editor.GFWorkspaceCommandStackImpl.execute(GFWorkspaceCommandStackImpl.java:128)
	at org.eclipse.emf.transaction.impl.AbstractTransactionalCommandStack.execute(AbstractTransactionalCommandStack.java:219)
	at org.eclipse.graphiti.internal.command.CommandExec.executeCommand(CommandExec.java:85)
	at org.eclipse.graphiti.ui.internal.command.ReconnectCommand.execute(ReconnectCommand.java:104)
	at org.eclipse.graphiti.ui.internal.editor.EmfOnGefCommand.execute(EmfOnGefCommand.java:54)
	at org.eclipse.graphiti.internal.command.GFPreparableCommand2.doExecute(GFPreparableCommand2.java:43)
	at org.eclipse.emf.transaction.RecordingCommand.execute(RecordingCommand.java:135)
	at org.eclipse.emf.workspace.EMFCommandOperation.doExecute(EMFCommandOperation.java:119)
	at org.eclipse.emf.workspace.AbstractEMFOperation.execute(AbstractEMFOperation.java:150)
	at org.eclipse.core.commands.operations.DefaultOperationHistory.execute(DefaultOperationHistory.java:516)
	at org.eclipse.emf.workspace.impl.WorkspaceCommandStackImpl.doExecute(WorkspaceCommandStackImpl.java:208)
	at org.eclipse.emf.transaction.impl.AbstractTransactionalCommandStack.execute(AbstractTransactionalCommandStack.java:165)
	at org.eclipse.graphiti.ui.internal.editor.GFWorkspaceCommandStackImpl.execute(GFWorkspaceCommandStackImpl.java:97)
	at org.eclipse.graphiti.ui.internal.editor.GFCommandStack.execute(GFCommandStack.java:138)
	at org.eclipse.gef.tools.AbstractTool.executeCommand(AbstractTool.java:425)
	at org.eclipse.gef.tools.AbstractTool.executeCurrentCommand(AbstractTool.java:438)
	at org.eclipse.gef.tools.ConnectionEndpointTracker.handleButtonUp(ConnectionEndpointTracker.java:172)
	at org.eclipse.graphiti.ui.internal.policy.GFConnectionEndpointTracker.handleButtonUp(GFConnectionEndpointTracker.java:46)
	at org.eclipse.gef.tools.AbstractTool.mouseUp(AbstractTool.java:1200)
	at org.eclipse.gef.tools.SelectionTool.mouseUp(SelectionTool.java:574)
	at org.eclipse.gef.EditDomain.mouseUp(EditDomain.java:301)
	at org.eclipse.gef.ui.parts.DomainEventDispatcher.dispatchMouseReleased(DomainEventDispatcher.java:380)
	at org.eclipse.draw2d.LightweightSystem$EventHandler.mouseUp(LightweightSystem.java:548)
	at org.eclipse.swt.widgets.TypedListener.handleEvent(TypedListener.java:220)
	at org.eclipse.swt.widgets.EventTable.sendEvent(EventTable.java:84)
	at org.eclipse.swt.widgets.Display.sendEvent(Display.java:4362)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1113)
	at org.eclipse.swt.widgets.Display.runDeferredEvents(Display.java:4180)
	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:3769)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine$4.run(PartRenderingEngine.java:1127)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:337)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine.run(PartRenderingEngine.java:1018)
	at org.eclipse.e4.ui.internal.workbench.E4Workbench.createAndRunUI(E4Workbench.java:156)
	at org.eclipse.ui.internal.Workbench$5.run(Workbench.java:694)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:337)
	at org.eclipse.ui.internal.Workbench.createAndRunWorkbench(Workbench.java:606)
	at org.eclipse.ui.PlatformUI.createAndRunWorkbench(PlatformUI.java:150)
	at org.eclipse.ui.internal.ide.application.IDEApplication.start(IDEApplication.java:139)
	at org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:196)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:134)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:104)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:380)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:235)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:669)
	at org.eclipse.equinox.launcher.Main.basicRun(Main.java:608)
	at org.eclipse.equinox.launcher.Main.run(Main.java:1515)
	at org.eclipse.equinox.launcher.Main.main(Main.java:1488)

**Comment by lhein** [comment]
testPR

**Comment by apupier** [comment]
+1

**Pull request merged**